### DISCUSSION TITLE
Gradient Computation with Inverse Wishart Prior

### DISCUSSION THREAD

**Discussion by higherorderderivatives2024** [meta-pytorch/botorch#1556] [Q&A]
Hi all, 

I am working with a multi-task Gaussian Process with a task matrix drawn from an Inverse Wishart prior, instead of the default LKJ Covariance prior. To implement this process, I am using the [newly implemented Inverse Wishart prior](https://github.com/cornellius-gp/gpytorch/blob/fd99baf587f6f50f414bb53b39a08a2f7d7a5e5b/gpytorch/priors/wishart_prior.py) in the latest, unstable version of gpytorch. 

After obtaining new data, I would like to fit this multi-task GP using the function fit_gpytorch_mll, but I am getting the following error:

RuntimeError: one of the variables needed for gradient computation has been modified by an inplace operation: [torch.DoubleTensor [1, 6]] is at version 1; expected version 0 instead. Hint: the backtrace further above shows the operation that failed to compute its gradient. The variable in question was changed in there or anywhere later. Good luck!

I don't run into this error if I use the default LKJ Covariance prior. One possibility is that there is something wrong in how I've drawn the task covariance matrix from the inverse wishart prior, and I've replicated my code here.

`train_Y_t = torch.transpose(train_Y,1,0)

 scale_matrix = data_covar_module(train_X).inv_matmul(train_Y, left_tensor = train_Y_t)

 task_covar_prior = InverseWishartPrior(2+train_X.shape[0], scale_matrix)`

Another possibility is that the Inverse Wishart prior in gpytorch could require modification to fix this. 

Any thoughts regarding this would be much appreciated - thanks a lot!

**Comment by Balandat** [comment]
Hmm interesting. This seems like an issue with the Wishart prior if the same code works with the LKJPrior works without any issues. I just took a look and I didn't see any obvious issues. 

Can you share a full code example to repro ? It could also be that some of the data that you're sticking into the prior or model was modified in place causing this error.

**Reply by higherorderderivatives2024** [reply]
Thank you! I will. However, I would like to clean up my code a bit first, and make sure there are no bugs on my side before I do.

In this spirit, do you think you could please suggest how I can implement the Kronecker MultiTask Gaussian Process with no observation noise (that is, no additional sigma^2*I term)?

I don't think I am implementing this the right way now.

**Reply by Balandat** [reply]
> how I can implement the Kronecker MultiTask Gaussian Process with no observation noise (that is, no additional sigma^2*I term)?

You should be able to just set `has_task_noise` and `has_global_noise` to `False` in the likelihood: https://github.com/cornellius-gp/gpytorch/blob/cda4bc54b773cb6a98979a95e9ec80310d198663/gpytorch/likelihoods/multitask_gaussian_likelihood.py#L190-L191

Thought this will likely cause a bunch of numerical issues when your Kernel is close to singular (e.g. if you have very similar data points etc.)

**Reply by higherorderderivatives2024** [reply]
I tried that- however, I run into the error that at least one of them has
to be true.

In the documentation for the likelihood, there is the following line:

if not has_task_noise and not has_global_noise:
  raise ValueError( "At least
  one of has_task_noise or has_global_noise must be specified. " "Attempting
  to specify a likelihood that has no noise terms." )

**Comment by higherorderderivatives2024** [comment]
To follow up on this, I used the suggested code for FixedTaskNoiseMultitaskLikelihood at the end of the following [discussion](https://github.com/cornellius-gp/gpytorch/issues/901), and ran into the following error message 

TypeError: torch.Size() takes an iterable of 'int' (item 0 is 'torch.Size') in the self.noise_covar() function in the likelihood.

I have attached my code, as well as the error [here](https://drive.google.com/file/d/1zw6Ixtk-moBNxCkg6tw4OvfNW2uOhR8I/view?usp=sharing).

Thanks a lot for your help!

**Reply by Balandat** [reply]
> TypeError: torch.Size() takes an iterable of 'int' (item 0 is 'torch.Size') in the self.noise_covar() function in the likelihood.

I don't see this error in the attached colab notebook, and I also can't produce it when running this on the latest botorch main branch (it runs, but slowly).
